app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt

```diff
@@ -0,0 +1,140 @@
+
+    notificationJob =
+        scope.launch {
+          while (true) {
+            checkNearbyUsersForCommonTags()
+            delay(CHECK_INTERVAL)
+          }
+        }
+  }
+
+  /** Stop monitoring for nearby users */
+  fun stopMonitoring() {
+    notificationJob?.cancel()
+    notificationJob = null
+  }
+
+  private fun isAppInForeground(): Boolean {
+    val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
+    val appProcesses = activityManager.runningAppProcesses ?: return false
+    val packageName = context.packageName
+
+    for (appProcess in appProcesses) {
+      if (appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND &&
+          appProcess.processName == packageName) {
+        return true
+      }
+    }
+    return false
+  }
+
+  private fun checkNearbyUsersForCommonTags() {
+    val selfProfile = profilesViewModel.selfProfile.value ?: return
+    val selfLocation = selfProfile.location ?: return
+
+    // Use the user's selected radius from FilterViewModel
+    profilesViewModel.getFilteredProfilesInRadius(
+        center = selfLocation,
+        radiusInMeters = filterViewModel.selectedRadius.value,
+        genders = filterViewModel.selectedGenders.value,
+        ageRange = filterViewModel.ageRange.value)
+
+    // Launch a coroutine to collect the filtered profiles
+    scope.launch {
+      // Only proceed if we are discoverable
+      if (!appDataStore.isDiscoverable.first()) return@launch
+
+      profilesViewModel.filteredProfiles.collectLatest { nearbyProfiles ->
+        // Only send notifications if the app is not in foreground
+        if (!isAppInForeground()) {
+          processNearbyProfiles(selfProfile, nearbyProfiles)
+        }
+      }
+    }
+  }
+
+  private fun processNearbyProfiles(selfProfile: Profile, nearbyProfiles: List<Profile>) {
+    val selfTags = selfProfile.tags
+
+    for (nearbyProfile in nearbyProfiles) {
+      if (nearbyProfile.uid == selfProfile.uid) continue // Skip self
+
+      // Find common tags
+      val commonTags = selfTags.intersect(nearbyProfile.tags.toSet())
+
+      if (commonTags.isNotEmpty()) {
+        // Send notification for the first common tag
+        val commonTag = commonTags.first()
+        try {
+          // Send notification to the other user
+          meetingRequestViewModel.engagementNotification(
+              targetToken = nearbyProfile.fcmToken ?: continue, tag = commonTag)
+
+          // Send notification to self as well
+          meetingRequestViewModel.engagementNotification(
+              targetToken = selfProfile.fcmToken ?: continue, tag = commonTag)
+        } catch (e: Exception) {
+          Log.e("EngagementNotification", "Failed to send notification", e)
+        }
+      }
+    }
+  }
+
+  // Add this method for testing purposes
+  @VisibleForTesting fun isMonitoring(): Boolean = notificationJob?.isActive == true
```

Comment:
This is very clever ! I didn't even now that you can make functions only visible for testing ! I think using this kind of function for testing in the futur